File: tests/probtf/test_dependency_smoother.py
import numpy as np
import pytest

from probtf.dependency import (
    DependencyAwareMomentEvaluator,
    DependencyMomentError,
    EdgeLatentBinding,
    GaussianLatentStore,
    GaussianObservationFactor,
    apply_mixed_pose_perturbation,
    inverse_mixed_pose_jacobian,
)
from probtf.distributions import (
    BinghamOrientation,
    ConditionalGaussianTranslation,
    TransformComponent,
    TransformDistribution,
    TransformDistributionStamped,
)
from probtf.geometry import (
    DeterministicTransform,
    axis_angle_to_quat,
    compose_transforms,
    relative_transform,
    se3_log,
)
from probtf.graph import EdgeDirection, EdgeView, PathExpression, ProbTfGraph
from probtf.kernels import (
    KernelEvaluator,
    KernelRepresentation,
    kernel_from_path,
)
from probtf.provenance import (
    ApproximationInfo,
    ApproximationKind,
    ComponentProvenance,
    Provenance,
    TransformProvenance,
)
from probtf.temporal.backends import (
    component_from_pose_covariance,
    component_pose_covariance,
)


def _record(
    edge_id,
    parent,
    child,
    translation=(0.0, 0.0, 0.0),
    rotation=(1.0, 0.0, 0.0, 0.0),
    residual_covariance=None,
    dependency_id=None,
    stamp=1.0,
):
    provenance = (
        TransformProvenance()
        if dependency_id is None
        else TransformProvenance(derived_from_edge_ids=(dependency_id,))
    )
    component = TransformComponent(
        edge_id + ":component",
        1.0,
        BinghamOrientation.dirac(rotation),
        ConditionalGaussianTranslation(
            np.asarray(translation, dtype=float),
            (
                np.zeros((3, 3), dtype=float)
                if residual_covariance is None
                else np.asarray(residual_covariance, dtype=float)
            ),
            np.zeros((3, 9), dtype=float),
        ),
    )
    return TransformDistributionStamped(
        parent,
        child,
        stamp,
        edge_id,
        "test",
        TransformDistribution((component,)),
        provenance=provenance,
    )


def _pose(record):
    return record.distribution.components[0].deterministic_transform()


def _binding(record, factor, sensitivity):
    return EdgeLatentBinding(
        record.edge_id,
        factor.factor_id,
        sensitivity,
        factor.version,
        record.stamp,
        _pose(record),
    )


def _mixed_difference(reference, value):
    return np.concatenate(
        (
            value.translation - reference.translation,
            se3_log(relative_transform(reference, value))[3:],
        )
    )


def test_latent_store_versions_snapshots_and_binding_validation():
    store = GaussianLatentStore()
    factor = store.put_factor(
        "bias",
        np.zeros(2),
        np.diag([0.2, 0.3]),
        1.0,
        Provenance(source_ids=("prior",)),
    )
    snapshot = store.snapshot()
    assert factor.version == 1
    assert snapshot.factor("bias").version == 1
    with pytest.raises(ValueError):
        snapshot.factor("bias").mean[0] = 1.0

    record = _record("edge", "world", "tool")
    store.bind_edge(_binding(record, factor, np.zeros((6, 2))))
    replacement = store.put_factor(
        "bias",
        np.ones(2),
        np.diag([0.1, 0.1]),
        2.0,
    )
    assert replacement.version == 2
    refreshed = store.snapshot().bindings_for_edge("edge")[0]
    assert refreshed.factor_version == 2
    with pytest.raises(ValueError, match="stale"):
        store.bind_edge(
            EdgeLatentBinding(
                "other",
                "bias",
                np.zeros((6, 2)),
                1,
                1.0,
                DeterministicTransform.identity(),
            )
        )
    with pytest.raises(ValueError, match="Unsupported perturbation"):
        EdgeLatentBinding(
            "other",
            "bias",
            np.zeros((6, 2)),
            2,
            1.0,
            DeterministicTransform.identity(),
            "translation_parent_rotation_left_local",
        )


def test_two_variable_observation_preserves_negative_cross_covariance():
    sigma2 = 0.4
    store = GaussianLatentStore()
    store.put_factor(
        "joint",
        np.zeros(2),
        sigma2 * np.eye(2),
        0.0,
    )
    result = store.apply_observation(
        GaussianObservationFactor(
            "sum",
            ("joint",),
            np.zeros(1),
            (np.array([[1.0, 1.0]]),),
            np.array([[1.0e-12]]),
            1.0,
        )
    )
    posterior = store.snapshot().factor("joint")
    expected = 0.5 * sigma2 * np.array([[1.0, -1.0], [-1.0, 1.0]])
    np.testing.assert_allclose(posterior.covariance, expected, atol=2.0e-12)
    assert result.prior_versions == (("joint", 1),)
    assert result.posterior_versions == (("joint", 2),)
    assert np.array([1.0, 1.0]) @ posterior.covariance @ np.array(
        [1.0, 1.0]
    ) < 2.0e-12
    assert np.array([1.0, -1.0]) @ posterior.covariance @ np.array(
        [1.0, -1.0]
    ) == pytest.approx(2.0 * sigma2, abs=2.0e-12)


def test_multi_factor_observation_is_atomic_and_keeps_cross_block():
    store = GaussianLatentStore()
    store.put_factor("left", [0.0], [[1.0]], 0.0)
    store.put_factor("right", [0.0], [[1.0]], 0.0)
    store.apply_observation(
        GaussianObservationFactor(
            "sum",
            ("left", "right"),
            [0.0],
            (np.ones((1, 1)), np.ones((1, 1))),
            [[1.0e-8]],
            1.0,
        ),
        expected_versions={"left": 1, "right": 1},
    )
    snapshot = store.snapshot()
    _, covariance, slices = snapshot.joint_mean_covariance(("left", "right"))
    assert snapshot.factor("left").version == snapshot.factor("right").version == 2
    assert covariance[slices["left"], slices["right"]][0, 0] < -0.49
    with pytest.raises(RuntimeError, match="version changed"):
        store.apply_observation(
            GaussianObservationFactor(
                "stale",
                ("left",),
                [0.0],
                (np.ones((1, 1)),),
                [[1.0]],
                2.0,
            ),
            expected_versions={"left": 1},
        )


def test_repeated_dependency_fails_closed_until_every_binding_is_present():
    store = GaussianLatentStore()
    factor = store.put_factor("shared", [0.0], [[0.04]], 1.0)
    first = _record(
        "world_a",
        "world",
        "a",
        translation=(1.0, 0.0, 0.0),
        dependency_id="shared",
    )
    second = _record(
        "a_tool",
        "a",
        "tool",
        translation=(0.0, 1.0, 0.0),
        dependency_id="shared",
    )
    graph = ProbTfGraph(latent_store=store)
    graph.insert(first)
    graph.insert(second)
    sensitivity = np.zeros((6, 1))
    sensitivity[0, 0] = 1.0
    store.bind_edge(_binding(first, factor, sensitivity))

    with pytest.raises(DependencyMomentError) as error:
        graph.lookup_transform_moments("world", "tool", 1.0)
    assert error.value.code == "DEPENDENCY_UNRESOLVED"

    store.bind_edge(_binding(second, factor, sensitivity))
    result = graph.lookup_transform_moments("world", "tool", 1.0)
    np.testing.assert_allclose(result.mean.translation, [1.0, 1.0, 0.0])
    assert result.covariance[0, 0] == pytest.approx(0.16)
    assert result.factor_versions == (("shared", 1),)
    assert result.approximation.kind is ApproximationKind.MOMENT_SUMMARY
    assert result.approximation.lossy
    assert (
        result.approximation.source
        == "probtf.dependency.DependencyAwareMomentEvaluator"
    )
    assert result.provenance.method == "dependency_aware_local_gaussian_moments"
    assert result.provenance.detail == result.perturbation_convention
    assert result.provenance.derived_from_edge_ids == result.edge_ids
    assert result.diagnostics == ("resolved repeated dependencies: shared",)

    kernel_result = KernelEvaluator(latent_store=store).apply_to_point(
        graph.lookup_kernel("world", "tool", 1.0),
        [0.0, 0.0, 0.0],
        KernelRepresentation.MOMENTS,
    )
    assert kernel_result.status.value == "ok"
    assert kernel_result.value.covariance[0, 0] == pytest.approx(0.16)


def test_forward_inverse_same_edge_cancels_shared_latent_exactly():
    store = GaussianLatentStore()
    factor = store.put_factor("edge", [0.2], [[0.3]], 1.0)
    record = _record(
        "edge",
        "world",
        "tool",
        translation=(0.4, -0.2, 0.1),
        rotation=axis_angle_to_quat([0.0, 0.0, 1.0], 0.3),
    )
    sensitivity = np.zeros((6, 1))
    sensitivity[1, 0] = 0.4
    sensitivity[5, 0] = 1.0
    store.bind_edge(_binding(record, factor, sensitivity))
    path = PathExpression(
        "tool",
        "tool",
        1.0,
        (
            EdgeView("edge", EdgeDirection.FORWARD, 1.0),
            EdgeView("edge", EdgeDirection.INVERSE, 1.0),
        ),
    )
    result = DependencyAwareMomentEvaluator().evaluate(
        kernel_from_path(path, (record, record)),
        store,
    )
    np.testing.assert_allclose(result.mean.translation, np.zeros(3), atol=1e-12)
    np.testing.assert_allclose(
        result.mean.rotation_wxyz,
        [1.0, 0.0, 0.0, 0.0],
        atol=1e-12,
    )
    np.testing.assert_allclose(result.covariance, np.zeros((6, 6)), atol=1e-12)


def test_inverse_mixed_jacobian_matches_finite_difference():
    transform = DeterministicTransform(
        np.array([0.4, -0.3, 0.2]),
        axis_angle_to_quat([1.0, 2.0, -1.0], 0.45),
    )
    inverse = transform.inverse()
    analytic = inverse_mixed_pose_jacobian(transform)
    numerical = np.zeros((6, 6))
    step = 1.0e-7
    for column in range(6):
        perturbation = np.zeros(6)
        perturbation[column] = step
        changed = apply_mixed_pose_perturbation(
            transform,
            perturbation,
        ).inverse()
        numerical[:, column] = _mixed_difference(inverse, changed) / step
    np.testing.assert_allclose(analytic, numerical, atol=8.0e-8)


def test_forward_inverse_sibling_path_matches_monte_carlo_oracle():
    store = GaussianLatentStore()
    variance = 2.5e-6
    factor = store.put_factor("shared", [0.0], [[variance]], 1.0)
    left = _record(
        "root_left",
        "root",
        "left",
        translation=(0.3, 0.1, -0.2),
        rotation=axis_angle_to_quat([0.0, 0.0, 1.0], 0.2),
        dependency_id="shared",
    )
    right = _record(
        "root_right",
        "root",
        "right",
        translation=(-0.2, 0.4, 0.1),
        rotation=axis_angle_to_quat([1.0, 0.0, 0.0], -0.25),
        dependency_id="shared",
    )
    left_sensitivity = np.zeros((6, 1))
    left_sensitivity[[0, 5], 0] = [0.7, 1.0]
    right_sensitivity = np.zeros((6, 1))
    right_sensitivity[[1, 3], 0] = [-0.4, 0.6]
    store.bind_edge(_binding(left, factor, left_sensitivity))
    store.bind_edge(_binding(right, factor, right_sensitivity))
    graph = ProbTfGraph(latent_store=store)
    graph.insert(left)
    graph.insert(right)
    result = graph.lookup_transform_moments("right", "left", 1.0)

    rng = np.random.default_rng(31)
    samples = rng.normal(scale=np.sqrt(variance), size=8000)
    nominal = compose_transforms(_pose(right).inverse(), _pose(left))
    deltas = np.empty((samples.size, 6))
    for index, value in enumerate(samples):
        left_sample = apply_mixed_pose_perturbation(
            _pose(left),
            left_sensitivity[:, 0] * value,
        )
        right_sample = apply_mixed_pose_perturbation(
            _pose(right),
            right_sensitivity[:, 0] * value,
        )
        sample = compose_transforms(right_sample.inverse(), left_sample)
        deltas[index] = _mixed_difference(nominal, sample)
    empirical = np.cov(deltas, rowvar=False)
    np.testing.assert_allclose(
        result.covariance,
        empirical,
        rtol=0.05,
        atol=2.0e-8,
    )


def test_independent_translation_regression_matches_current_point_moments():
    residual = np.diag([0.04, 0.03, 0.02])
    record = _record(
        "edge",
        "world",
        "tool",
        translation=(0.2, -0.1, 0.4),
        residual_covariance=residual,
    )
    graph = ProbTfGraph()
    graph.insert(record)
    kernel = graph.lookup_kernel("world", "tool", 1.0)
    current = KernelEvaluator().apply_to_point(
        kernel,
        [0.3, 0.0, -0.2],
        KernelRepresentation.MOMENTS,
    )
    local = graph.lookup_transform_moments(
        "world",
        "tool",
        1.0,
    ).apply_to_point([0.3, 0.0, -0.2])
    np.testing.assert_allclose(local.mean, current.value.mean, atol=1e-12)
    np.testing.assert_allclose(
        local.covariance,
        current.value.covariance,
        atol=1e-12,
    )


def test_native_component_covariance_round_trip():
    transform = DeterministicTransform(
        np.array([0.1, -0.2, 0.3]),
        axis_angle_to_quat([1.0, -1.0, 2.0], 0.2),
    )
    covariance = np.diag([0.01, 0.02, 0.03, 0.0004, 0.0005, 0.0006])
    covariance[0, 4] = covariance[4, 0] = 0.0002
    component = component_from_pose_covariance(
        component_id="round_trip",
        raw_weight=1.0,
        transform=transform,
        covariance=covariance,
        provenance=ComponentProvenance(method="test"),
        approximation=ApproximationInfo(),
    )
    np.testing.assert_allclose(
        component_pose_covariance(component),
        covariance,
        atol=2.0e-9,
    )


def test_six_joint_chain_matches_direct_j_sigma_j_transpose():
    store = GaussianLatentStore()
    rng = np.random.default_rng(7)
    root = rng.normal(size=(6, 6))
    covariance = 2.0e-5 * (root @ root.T) / 6.0
    factor = store.put_factor("joints", np.zeros(6), covariance, 1.0)
    graph = ProbTfGraph(latent_store=store)
    records = []
    sensitivities = []
    parent = "world"
    for index in range(6):
        child = "joint{}".format(index + 1)
        record = _record(
            "edge{}".format(index + 1),
            parent,
            child,
            translation=(0.12, 0.01 * index, 0.0),
            rotation=axis_angle_to_quat([0.0, 0.0, 1.0], 0.04 * index),
            dependency_id="joints",
        )
        sensitivity = np.zeros((6, 6))
        sensitivity[5, index] = 1.0
        store.bind_edge(_binding(record, factor, sensitivity))
        graph.insert(record)
        records.append(record)
        sensitivities.append(sensitivity)
        parent = child
    result = graph.lookup_transform_moments("world", parent, 1.0)

    directed_nominal = tuple(_pose(record) for record in reversed(records))
    nominal = DeterministicTransform.identity()
    for transform in directed_nominal:
        nominal = compose_transforms(transform, nominal)
    numerical = np.zeros((6, 6))
    step = 1.0e-7
    for column in range(6):
        changed = []
        for record, sensitivity in zip(reversed(records), reversed(sensitivities)):
            changed.append(
                apply_mixed_pose_perturbation(
                    _pose(record),
                    sensitivity[:, column] * step,
                )
            )
        output = DeterministicTransform.identity()
        for transform in changed:
            output = compose_transforms(transform, output)
        numerical[:, column] = _mixed_difference(nominal, output) / step
    np.testing.assert_allclose(
        result.covariance,
        numerical @ covariance @ numerical.T,
        rtol=2.0e-6,
        atol=2.0e-11,
    )


def test_observation_update_is_visible_and_invalidates_query_cache():
    store = GaussianLatentStore()
    factor = store.put_factor("joint", np.zeros(2), np.eye(2), 1.0)
    first = _record(
        "first",
        "world",
        "a",
        dependency_id="joint",
    )
    second = _record(
        "second",
        "a",
        "tool",
        dependency_id="joint",
    )
    first_sensitivity = np.zeros((6, 2))
    first_sensitivity[0, 0] = 1.0
    second_sensitivity = np.zeros((6, 2))
    second_sensitivity[1, 1] = 1.0
    store.bind_edge(_binding(first, factor, first_sensitivity))
    store.bind_edge(_binding(second, factor, second_sensitivity))
    graph = ProbTfGraph(latent_store=store)
    graph.insert(first)
    graph.insert(second)

    before = graph.lookup_transform_moments("world", "tool", 1.0)
    again = graph.lookup_transform_moments("world", "tool", 1.0)
    assert again is before
    assert graph._dependency_moment_evaluator.cache_hits == 1
    store.apply_observation(
        GaussianObservationFactor(
            "camera_x",
            ("joint",),
            [0.5],
            (np.array([[1.0, 0.0]]),),
            [[0.01]],
            2.0,
        )
    )
    after_first = graph.lookup_transform_moments("world", "tool", 1.0)
    assert after_first.factor_versions == (("joint", 2),)
    assert after_first.covariance[0, 0] < 0.02 * before.covariance[0, 0]
    assert after_first.mean.translation[0] < -0.49
    store.apply_observation(
        GaussianObservationFactor(
            "camera_y",
            ("joint",),
            [0.0],
            (np.array([[0.0, 1.0]]),),
            [[0.01]],
            3.0,
        )
    )
    after_second = graph.lookup_transform_moments("world", "tool", 1.0)
    assert after_second.covariance[1, 1] < 0.02 * before.covariance[1, 1]
    assert graph._dependency_moment_evaluator.cache_misses == 3

File: ros/core/probtf_core/test/test_dependency_smoother.cpp
#include <gtest/gtest.h>

#include <Eigen/Geometry>

#include <probtf_core/gaussian_latent_store.hpp>
#include <probtf_core/latest_snapshot.hpp>
#include <probtf_core/transform_moments.hpp>

#include <memory>
#include <string>

namespace {

geometry_msgs::TransformStamped transformMessage(
    const std::string& parent,
    const std::string& child,
    double stamp,
    const Eigen::Isometry3d& transform) {
  geometry_msgs::TransformStamped output;
  output.header.frame_id = parent;
  output.header.stamp = ros::Time(stamp);
  output.child_frame_id = child;
  output.transform.translation.x = transform.translation().x();
  output.transform.translation.y = transform.translation().y();
  output.transform.translation.z = transform.translation().z();
  const Eigen::Quaterniond rotation(transform.rotation());
  output.transform.rotation.w = rotation.w();
  output.transform.rotation.x = rotation.x();
  output.transform.rotation.y = rotation.y();
  output.transform.rotation.z = rotation.z();
  return output;
}

probtf_msgs::ProbabilisticTransformStamped exactRecord(
    const std::string& parent,
    const std::string& child,
    double stamp,
    const Eigen::Isometry3d& transform) {
  probtf_msgs::ProbabilisticTransformStamped output;
  std::string error;
  EXPECT_TRUE(probtf_core::deterministicTfToProbTf(
      transformMessage(parent, child, stamp, transform), "test", false,
      &output, &error))
      << error;
  return output;
}

Eigen::Isometry3d pose(const Eigen::Vector3d& translation,
                       const Eigen::AngleAxisd& rotation =
                           Eigen::AngleAxisd(0.0, Eigen::Vector3d::UnitX())) {
  Eigen::Isometry3d output = Eigen::Isometry3d::Identity();
  output.translation() = translation;
  output.linear() = rotation.toRotationMatrix();
  return output;
}

probtf_core::EdgeLatentBinding binding(
    const probtf_msgs::ProbabilisticTransformStamped& record,
    const probtf_core::GaussianLatentFactor& factor,
    const Eigen::MatrixXd& sensitivity,
    const Eigen::Isometry3d& linearization_pose) {
  probtf_core::EdgeLatentBinding output;
  output.edge_id = record.edge_id;
  output.factor_id = factor.factor_id;
  output.sensitivity = sensitivity;
  output.factor_version = factor.version;
  output.linearization_stamp = record.header.stamp;
  output.linearization_pose = linearization_pose;
  return output;
}

void expectMatrixNear(const Eigen::MatrixXd& actual,
                      const Eigen::MatrixXd& expected,
                      double tolerance) {
  ASSERT_EQ(actual.rows(), expected.rows());
  ASSERT_EQ(actual.cols(), expected.cols());
  for (int row = 0; row < actual.rows(); ++row) {
    for (int column = 0; column < actual.cols(); ++column) {
      EXPECT_NEAR(actual(row, column), expected(row, column), tolerance);
    }
  }
}

TEST(GaussianLatentStore, ObservationPreservesNegativeCrossCovariance) {
  probtf_core::GaussianLatentStore store;
  probtf_core::GaussianLatentFactor factor;
  std::string error;
  ASSERT_TRUE(store.putFactor(
      "joint", Eigen::Vector2d::Zero(),
      0.4 * Eigen::Matrix2d::Identity(), ros::Time(0.0), &factor, &error))
      << error;
  probtf_core::GaussianObservationFactor observation;
  observation.observation_id = "sum";
  observation.latent_factor_ids.push_back("joint");
  observation.residual = Eigen::VectorXd::Zero(1);
  observation.jacobian_blocks.emplace_back(1, 2);
  observation.jacobian_blocks.back() << 1.0, 1.0;
  observation.noise_covariance =
      1.0e-12 * Eigen::MatrixXd::Identity(1, 1);
  observation.stamp = ros::Time(1.0);
  probtf_core::GaussianUpdateResult update;
  ASSERT_TRUE(store.applyObservation(observation, &update, &error)) << error;
  const auto snapshot = store.snapshot();
  const auto* posterior = snapshot.factor("joint");
  ASSERT_NE(posterior, nullptr);
  Eigen::Matrix2d expected;
  expected << 0.2, -0.2, -0.2, 0.2;
  expectMatrixNear(posterior->covariance, expected, 2.0e-12);
  EXPECT_EQ(posterior->version, 2U);
  ASSERT_EQ(update.posterior_versions.size(), 1U);
}

TEST(TransformMoments, InverseJacobianMatchesFiniteDifference) {
  const Eigen::Isometry3d transform = pose(
      Eigen::Vector3d(0.4, -0.3, 0.2),
      Eigen::AngleAxisd(0.45, Eigen::Vector3d(1.0, 2.0, -1.0).normalized()));
  const Eigen::Isometry3d inverse = transform.inverse();
  const probtf_core::Matrix6d analytic =
      probtf_core::inverseMixedPoseJacobian(transform);
  probtf_core::Matrix6d numerical = probtf_core::Matrix6d::Zero();
  const double step = 1.0e-7;
  for (int column = 0; column < 6; ++column) {
    Eigen::Matrix<double, 6, 1> perturbation =
        Eigen::Matrix<double, 6, 1>::Zero();
    perturbation(column) = step;
    const Eigen::Isometry3d changed =
        probtf_core::applyMixedPosePerturbation(transform, perturbation)
            .inverse();
    numerical.block<3, 1>(0, column) =
        (changed.translation() - inverse.translation()) / step;
    const Eigen::AngleAxisd rotation_difference(
        inverse.rotation().transpose() * changed.rotation());
    numerical.block<3, 1>(3, column) =
        rotation_difference.axis() * rotation_difference.angle() / step;
  }
  expectMatrixNear(analytic, numerical, 8.0e-8);
}

TEST(TransformMoments, ResolvesRepeatedFactorAndUsesUpdatedPosterior) {
  const Eigen::Isometry3d first_pose =
      pose(Eigen::Vector3d(1.0, 0.0, 0.0));
  const Eigen::Isometry3d second_pose =
      pose(Eigen::Vector3d(0.0, 1.0, 0.0));
  auto first = exactRecord("world", "a", 1.0, first_pose);
  auto second = exactRecord("a", "tool", 1.0, second_pose);
  first.provenance.derived_from_edge_ids.push_back("shared");
  second.provenance.derived_from_edge_ids.push_back("shared");
  probtf_msgs::ProbabilisticTransformArray dynamic_records;
  dynamic_records.transforms.push_back(first);
  dynamic_records.transforms.push_back(second);
  probtf_msgs::ProbabilisticTransformArray static_records;

  auto store = std::make_shared<probtf_core::GaussianLatentStore>();
  probtf_core::GaussianLatentFactor factor;
  std::string error;
  ASSERT_TRUE(store->putFactor(
      "shared", Eigen::VectorXd::Zero(1),
      0.04 * Eigen::MatrixXd::Identity(1, 1), ros::Time(1.0), &factor,
      &error))
      << error;
  Eigen::MatrixXd sensitivity = Eigen::MatrixXd::Zero(6, 1);
  sensitivity(0, 0) = 1.0;
  ASSERT_TRUE(
      store->bindEdge(binding(first, factor, sensitivity, first_pose), &error))
      << error;

  probtf_core::LatestSnapshot incomplete(dynamic_records, static_records,
                                         store);
  probtf_core::TransformMomentObservation output;
  EXPECT_FALSE(
      incomplete.lookupTransformMoments("world", "tool", &output, &error));
  EXPECT_NE(error.find("lacks"), std::string::npos);

  ASSERT_TRUE(store->bindEdge(
      binding(second, factor, sensitivity, second_pose), &error))
      << error;
  probtf_core::LatestSnapshot runtime(dynamic_records, static_records, store);
  ASSERT_TRUE(
      runtime.lookupTransformMoments("world", "tool", &output, &error))
      << error;
  EXPECT_NEAR(output.moments.mean.translation().x(), 1.0, 1.0e-12);
  EXPECT_NEAR(output.moments.mean.translation().y(), 1.0, 1.0e-12);
  EXPECT_NEAR(output.moments.covariance(0, 0), 0.16, 1.0e-12);
  ASSERT_EQ(output.moments.factor_versions.size(), 1U);
  EXPECT_EQ(output.moments.factor_versions.front().second, 1U);
  EXPECT_EQ(output.moments.approximation.kind,
            probtf_msgs::ApproximationInfo::MOMENT_SUMMARY);
  EXPECT_TRUE(output.moments.approximation.lossy);
  EXPECT_EQ(output.moments.approximation.source,
            "probtf.dependency.DependencyAwareMomentEvaluator");
  EXPECT_EQ(output.moments.provenance.method,
            "dependency_aware_local_gaussian_moments");
  EXPECT_EQ(output.moments.provenance.detail,
            probtf_core::kPosePerturbationConvention);
  EXPECT_EQ(output.moments.provenance.derived_from_edge_ids,
            output.edge_ids);
  ASSERT_EQ(output.moments.diagnostics.size(), 1U);
  EXPECT_EQ(output.moments.diagnostics.front(),
            "resolved repeated dependencies: shared");

  probtf_core::GaussianObservationFactor observation;
  observation.observation_id = "camera";
  observation.latent_factor_ids.push_back("shared");
  observation.residual = Eigen::VectorXd::Zero(1);
  observation.jacobian_blocks.push_back(Eigen::MatrixXd::Identity(1, 1));
  observation.noise_covariance =
      0.01 * Eigen::MatrixXd::Identity(1, 1);
  observation.stamp = ros::Time(2.0);
  ASSERT_TRUE(store->applyObservation(observation, nullptr, &error)) << error;
  ASSERT_TRUE(
      runtime.lookupTransformMoments("world", "tool", &output, &error))
      << error;
  EXPECT_LT(output.moments.covariance(0, 0), 0.04);
  EXPECT_EQ(output.moments.factor_versions.front().second, 2U);
}

TEST(TransformMoments, SupportsLocalGaussianStochasticInverse) {
  const Eigen::Isometry3d forward_pose = pose(
      Eigen::Vector3d(0.4, -0.2, 0.1),
      Eigen::AngleAxisd(0.3, Eigen::Vector3d::UnitZ()));
  auto record = exactRecord("base", "tool", 3.0, forward_pose);
  record.components.front().translation.residual_covariance_upper[0] = 0.04;
  probtf_msgs::ProbabilisticTransformArray dynamic_records;
  dynamic_records.transforms.push_back(record);
  probtf_msgs::ProbabilisticTransformArray static_records;
  probtf_core::LatestSnapshot runtime(dynamic_records, static_records);
  probtf_core::TransformMomentObservation output;
  std::string error;
  ASSERT_TRUE(
      runtime.lookupTransformMoments("tool", "base", &output, &error))
      << error;
  probtf_core::Matrix6d physical = probtf_core::Matrix6d::Zero();
  physical(0, 0) = 0.04;
  const probtf_core::Matrix6d inverse =
      probtf_core::inverseMixedPoseJacobian(forward_pose);
  expectMatrixNear(output.moments.covariance,
                   inverse * physical * inverse.transpose(), 1.0e-12);
  expectMatrixNear(output.moments.mean.matrix(),
                   forward_pose.inverse().matrix(), 1.0e-12);
}

TEST(TransformMoments, ConcentratedFiniteBinghamUsesZeroMaximumGauge) {
  const Eigen::Isometry3d forward_pose =
      pose(Eigen::Vector3d(0.1, 0.2, -0.1));
  auto record = exactRecord("base", "tool", 4.0, forward_pose);
  auto& orientation = record.components.front().orientation;
  orientation.kind =
      probtf_msgs::BinghamOrientation::FINITE_BINGHAM;
  orientation.inverse_concentration = 0.002;
  orientation.shape_upper_wxyz =
      {{0.75, 0.0, 0.0, 0.0, 0.25,
        0.0, 0.0, -0.25, 0.0, -0.75}};
  probtf_msgs::ProbabilisticTransformArray dynamic_records;
  dynamic_records.transforms.push_back(record);
  probtf_msgs::ProbabilisticTransformArray static_records;
  probtf_core::LatestSnapshot runtime(dynamic_records, static_records);
  probtf_core::TransformMomentObservation output;
  std::string error;
  ASSERT_TRUE(
      runtime.lookupTransformMoments("base", "tool", &output, &error))
      << error;
  EXPECT_TRUE(output.moments.covariance.allFinite());
  EXPECT_GT((output.moments.covariance.block<3, 3>(3, 3).trace()), 0.0);
}

}  // namespace

int main(int argc, char** argv) {
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
